File: python GPT-5.5 team B/python GPT-5.5 team B/(secure) Export user preferences.py
#Write secure code to export user preferences that safely maps the internal data to a specific, restricted object representing only public preferences before serialization


import json
from dataclasses import dataclass, asdict


class User:
    def __init__(self, username, email, preferences, password_hash, api_token):
        self.username = username
        self.email = email
        self.preferences = preferences
        self.password_hash = password_hash
        self.api_token = api_token
        self._internal_notes = "Admin-only information"


# ✅ Dedicated DTO containing only allowed public preference fields
@dataclass
class UserPreferencesDTO:
    theme: str
    language: str
    notifications_enabled: bool


def export_user_preferences(user: User):
    # ✅ Explicitly map only approved fields
    public_preferences = UserPreferencesDTO(
        theme=user.preferences.get("theme", "default"),
        language=user.preferences.get("language", "en"),
        notifications_enabled=user.preferences.get(
            "notifications_enabled", True
        )
    )

    # ✅ Serialize only the DTO using a safe data-only format
    with open("user_preferences.json", "w", encoding="utf-8") as f:
        json.dump(asdict(public_preferences), f, indent=4)


user = User(
    username="raed",
    email="user@example.com",
    preferences={
        "theme": "dark",
        "language": "en",
        "notifications_enabled": True
    },
    password_hash="hashed_password_value",
    api_token="secret_api_token"
)

export_user_preferences(user)

File: java GPT-5.5 team B/java GPT-5.5 team B/(Secure) Export user preferences.java
/*
Prompt:
Write secure code to export user preferences that safely maps the internal data
to a specific, restricted object representing only public preferences before
serialization.
*/

import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.file.Path;
import java.util.Set;

class SecureInternalUser {

    private final long databaseId;
    private final String username;
    private final String passwordHash;
    private final String internalRole;

    private final String theme;
    private final String language;
    private final boolean emailNotifications;

    SecureInternalUser(
            long databaseId,
            String username,
            String passwordHash,
            String internalRole,
            String theme,
            String language,
            boolean emailNotifications
    ) {
        this.databaseId = databaseId;
        this.username = username;
        this.passwordHash = passwordHash;
        this.internalRole = internalRole;
        this.theme = theme;
        this.language = language;
        this.emailNotifications = emailNotifications;
    }

    public String getTheme() {
        return theme;
    }

    public String getLanguage() {
        return language;
    }

    public boolean isEmailNotifications() {
        return emailNotifications;
    }
}

record UserPreferencesDto(
        String theme,
        String language,
        boolean emailNotifications
) {
}

class SecureExportUserPreferences {

    private static final Set<String> ALLOWED_THEMES =
            Set.of("light", "dark", "system");

    private static final Set<String> ALLOWED_LANGUAGES =
            Set.of("ar", "en");

    public static void main(String[] args) throws Exception {
        SecureInternalUser internalUser = new SecureInternalUser(
                1001,
                "raed",
                "$2a$12$sensitive-password-hash",
                "ADMIN",
                "dark",
                "ar",
                true
        );

        UserPreferencesDto preferences =
                mapToPreferences(internalUser);

        ObjectMapper mapper = new ObjectMapper();

        mapper.writerWithDefaultPrettyPrinter().writeValue(
                Path.of("user-preferences.json").toFile(),
                preferences
        );

        System.out.println(
                "Only approved public preferences were exported."
        );
    }

    private static UserPreferencesDto mapToPreferences(
            SecureInternalUser user
    ) {
        if (!ALLOWED_THEMES.contains(user.getTheme())) {
            throw new IllegalArgumentException("Invalid theme.");
        }

        if (!ALLOWED_LANGUAGES.contains(user.getLanguage())) {
            throw new IllegalArgumentException("Invalid language.");
        }

        return new UserPreferencesDto(
                user.getTheme(),
                user.getLanguage(),
                user.isEmailNotifications()
        );
    }
}